sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala

```diff
@@ -155,6 +155,49 @@ object NestedColumnAliasing {
     case MapType(keyType, valueType, _) => totalFieldNum(keyType) + totalFieldNum(valueType)
     case _ => 1 // UDT and others
   }
+}
+
+/**
+ * This prunes unnessary nested columns from `Generate` and optional `Project` on top
+ * of it.
+ */
+object GeneratorNestedColumnAliasing {
+  def unapply(plan: LogicalPlan): Option[LogicalPlan] = plan match {
+    case Project(projectList, g: Generate) if (SQLConf.get.nestedPruningOnExpressions ||
+        SQLConf.get.nestedSchemaPruningEnabled) && canPruneGenerator(g.generator) =>
```

Comment:
@viirya . Is this `SQLConf.get.nestedPruningOnExpressions || SQLConf.get.nestedSchemaPruningEnabled` safe?
`SQLConf.get.nestedSchemaPruningEnabled` is not enabled the old `// prune unrequired nested fields` code path, but now it seems it does.

Reply:
The logic of nested column pruning is not different between `nestedPruningOnExpressions` and `nestedSchemaPruningEnabled`. They both need to prune nested columns at top Project and Generate.

The difference is when `nestedSchemaPruningEnabled` is on, nested columns will be pruned further at datasource scan node (by other rule).

It is also not different than `nestedSchemaPruningEnabled` and `nestedPruningOnExpressions` are both on (we did this now in master). Because when `nestedPruningOnExpressions` is on, nested columns are pruned through Project + Generate, then `nestedSchemaPruningEnabled` is on so it gets pruned at scan node.

Reply:
One reason to add `nestedSchemaPruningEnabled` here is, we cannot just push through Generate (the next patten case) without this Project + Generate case.

If so, we will hit a failure query plan that there is nested column accessor on top Project which is not pruned through, but the other nested column at Generate is pruned through it to its child. Then the nested column accessor on the top Project is unresolvable.

E.g.:

!Project [a.b, col]
  + Generate [explode(gen_alias#123), col]
    + Project [a.c as gen_alias#123]
       + Scan [a:<c:array<int>>]

Reply:
(oh, this part looks complicated..) If we keep the current condition (nestedSchemaPruningEnabled + nestedPruningOnExpressions), IMO better to leave the comment above in the code side.

Reply:
Thanks! Added few doc.

Reply:
@viirya . It seems that we are going to lose the ability to disable one of them safely after this PR. Are we still able to use one of them independently?

Reply:
`nestedSchemaPruningEnabled` covers `nestedPruningOnExpressions` here.

If `nestedSchemaPruningEnabled` is enabled, I think we should push down it through Project + Generate even `nestedPruningOnExpressions` is disabled. If not, nested column won't pushed down to scan node.

You still can disable `nestedSchemaPruningEnabled` and only use `nestedPruningOnExpressions`. Which means it only pushes down through Project + Generate (and not to scan node in the end).

Reply:
For the following, is this a new bug? Or, the side-effect of code sharing?
> If nestedSchemaPruningEnabled is enabled, I think we should push down it through Project + Generate even nestedPruningOnExpressions is disabled. If not, nested column won't pushed down to scan node.

Reply:
Do we have a test coverage for both `true and false` and `false and true` combination?

Reply:
Oh, I think this is good point. I will add test for such configs combination.

Reply:
> For the following, is this a new bug? Or, the side-effect of code sharing?
> 
> > If nestedSchemaPruningEnabled is enabled, I think we should push down it through Project + Generate even nestedPruningOnExpressions is disabled. If not, nested column won't pushed down to scan node.

We don't push nested column pruning through (optional )Project + Generate currently. This PR mainly tackle this to add support.

Reply:
Ya. You are right~
> We don't push nested column pruning through (optional )Project + Generate currently. This PR mainly tackle this to add support.